Method parity between vanilla registries and Forge registry wrappers

dhyces opened the pull request:
Fixes #8758. This changes the implementation of `getOrCreateHolder` to more closely match the vanilla implementation (it can return an errored DataResult in cases where the registry value is needed or registry is frozen, rather than throwing) and moves the original implementation to `getOrCreateHolderOrThrow`. We also add the method `getResourceKeys` to IForgeRegistry. This also extends the relevant methods in NamespacedWrapper and NamespacedDefaultedWrapper.

As a side-note, I had a case before where I mistakenly did not call `ResourceKey#location` and when the method returned the errored DataResult, the error did not happen to have much helpful information regarding what the key and registry was. Perhaps when the holderLookup field is nonnull it should still say what key was attempted in both "getOrCreateHolder..." methods?

Commit:
getOrCreateHolderOrThrow impl

-Effectively moved what was previously getOrCreateHolder to the new getOrCreateHolderOrThrow
-Rewrote getOrCreateHolder to return a DataResult
-Change the getOrCreateHolder methods in NamespacedWrapper and NamespacedDefaultedWrapper
-Extend getOrCreateHolderOrThrow in NamespacedWrapper and NamespacedDefaultedWrapper

Commit:
Fixed missed method call

Commit:
Fixed missed method call in "getOrCreateHolderOrThrow impl"

Commit:
Merge branch 'registry-method-parity' of https://github.com/dhyces/MinecraftForge into registry-method-parity

Commit:
registryKeySet impl

-Added getResourceKeys to IForgeRegistry
-Implemented getResourceKeys in ForgeRegistry
-Extended registryKeySet in both NamespacedWrapper and NamespacedDefaultedWrapper

The name "getResourceKeys" was going to be "getRegistryKeys" initially, though it would clash heavily with "getRegistryKey". The methods "getResourceKey" and "getResourceKeys" are syntactically different, where one accepts a parameter and the other does not.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=8766) <br/>All committers have signed the CLA.

SizableShrimp commented on src/main/java/net/minecraftforge/registries/IForgeRegistry.java:
```diff
@@ -46,6 +46,7 @@
     @NotNull Optional<ResourceKey<V>> getResourceKey(V value);
 
     @NotNull Set<ResourceLocation>         getKeys();
+    @NotNull Set<ResourceKey<V>>           getResourceKeys();
```
Exposing this is not necessarily for functionality. Please delete this from the interface, and make the relevant method in ForgeRegistry package-private.

SizableShrimp commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -88,7 +89,24 @@ Optional<Holder<T>> getHolder(T value)
+        if (ref == null)
+        {
+            if (this.holderLookup != null)
+                return DataResult.error("This registry can't create new holders without value");
```
Can you add `" (requested key" + key + ")"` to the end of this error message like vanilla?
This additional context would also probably be good in the same error message in `getOrCreateHolderOrThrow`

Commit:
Made requested changes

-removed IForgeRegistry interface getResourceKeys method
-made getResourceKeys package-private in ForgeRegistry
-added ResourceKey info to errors when holderLookup is non-null

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -88,12 +89,29 @@ Optional<Holder<T>> getHolder(T value)
+        if (ref == null)
+        {
+            if (this.holderLookup != null)
+                return DataResult.error("This registry can't create new holders without value. (requested key" + key + ")");
```
Small detail:
```suggestion
                return DataResult.error("This registry can't create new holders without value (requested key: " + key + ")");
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -88,12 +89,29 @@ Optional<Holder<T>> getHolder(T value)
         {
             if (this.holderLookup != null)
-                throw new IllegalStateException("This registry can't create new holders without value");
+                throw new IllegalStateException("This registry can't create new holders without value. (requested key" + key + ")");
```
Small detail:
```suggestion
                throw new IllegalStateException("This registry can't create new holders without value. (requested key: " + key + ")");
```

sciwhiz12 reviewed (changes requested):
Also, please add `fixes #8758` to your PR description so this PR is linked to that issue automatically by GitHub.

Commit:
Small error formatting changes

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -88,12 +89,29 @@ Optional<Holder<T>> getHolder(T value)
+    }
+
+    Holder<T> getOrCreateHolderOrThrow(ResourceKey<T> key)
     {
         return this.holdersByName.computeIfAbsent(key.location(), k ->
         {
             if (this.holderLookup != null)
-                throw new IllegalStateException("This registry can't create new holders without value");
+                throw new IllegalStateException("This registry can't create new holders without value (requested key: " + key + ")");
             if (this.frozen)
                 throw new IllegalStateException("Registry is already frozen (trying to add key " + k + ")");
             return Holder.Reference.createStandAlone(this.self, key);
```
Could we reimplement this method to instead call `getOrCreateHolder` and unwrap the given `DataResult`, turning an error message into an `IllegalStateException` (perhaps using `LambdaExceptionUtils#rethrowConsumer` to bypass the pesky pain of checked exceptions)?

sciwhiz12 replied:
I've been advised that the correct lambda exception utility class to suggest is <code>cpw.mods.modlauncher.api.Lam<i>db</i>aExceptionUtils</code> and **not** `cpw.mods.util.LambdaExceptionUtils` -- yes, the one with the misspell of `Lamdba`. Apparently, the one under `cpw.mods.util` is from securejarhandler, which does not export that package, causing an illegal access error during runtime.

dhyces replied:
Right, so, there is a method in DataResult called "getOrThrow" that throws a RuntimeException. Should we use that and have a RuntimeException or have a more specific IllegalStateException?

SizableShrimp replied:
I'd prefer to keep IllegalStateException

sciwhiz12 reviewed (changes requested):
The new method looks fine to me, but there's deduplication of code that can be done here. (Also, please rerequest my review when you've applied your changes. @SizableShrimp can't be rerequested for review because of GitHub limitations.)

Commit:
Rewrite getOrCreateHolderOrThrow - 1

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java:
```diff
@@ -88,15 +90,31 @@ Optional<Holder<T>> getHolder(T value)
+        return DataResult.success(ref);
+    }
+
+    Holder<T> getOrCreateHolderOrThrow(ResourceKey<T> key)
+    {
+        return LamdbaExceptionUtils.uncheck(() ->
+        {
+            DataResult<Holder<T>> dataResult = getOrCreateHolder(key);
+            if (dataResult.error().isPresent())
+                throw new IllegalStateException(dataResult.error().get().message());
+            return dataResult.result().get();
```
You may have misunderstood what I meant by using `LamdbaExceptionUtils`. In this form,  `LamdbaExceptionUtils#uncheck` effectively does nothing as `IllegalStateException` is already an unchecked exception. What I meant to suggest was to call `DataResult#getOrThrow` like so:
```java
return getCreateHolder(key).getOrThrow(false, LamdbaExceptionUtils.rethrowConsumer((String message) -> {
    throw new IllegalStateException(message);
}));
```

SizableShrimp replied:
Is it even necessary in this form?

sciwhiz12 replied:
Actually, that's a fair point which I hadn't considered 👀 -- we don't even need `LamdbaExceptionUtils` because `IllegalStateException` is an unchecked exception.

dhyces replied:
Yes, unfortunately, misunderstood its use. Figured we only wanted the illegal state, but I suppose it doesn't impact anything if we throw both. I'll push the revision now.

Commit:
Revise getOrCreateHolderOrThrow - 2
